x-pack/legacy/plugins/ml/public/application/explorer/explorer_directive.tsx

```diff
@@ -0,0 +1,120 @@
+
+import { interval$ } from '../components/controls/select_interval';
+import { severity$ } from '../components/controls/select_severity';
+import { showCharts$ } from '../components/controls/checkbox_showcharts';
+import { subscribeAppStateToObservable } from '../util/app_state_utils';
+
+import { Explorer } from './explorer';
+import { EXPLORER_ACTION } from './explorer_constants';
+import { explorerAction$, explorerAppState$ } from './explorer_dashboard_service';
+import { getExplorerDefaultAppState } from './reducers';
+
+interface ExplorerScope extends IScope {
+  appState: IAppState;
+}
+
+module.directive('mlAnomalyExplorer', function(
+  globalState: State,
+  $rootScope: IRootScopeService,
+  AppState: AppStateClass
+) {
+  function link($scope: ExplorerScope, element: IRootElementService) {
+    const subscriptions = new Subscription();
+
+    const { jobSelectService$, unsubscribeFromGlobalState } = jobSelectServiceFactory(globalState);
+
+    ReactDOM.render(
+      <I18nContext>
+        <Explorer
+          {...{
+            globalState,
+            jobSelectService$,
+          }}
+        />
+      </I18nContext>,
+      element[0]
+    );
+
+    // Initialize the AppState in which to store swimlane and filter settings.
+    // AppState is used to store state in the URL.
+    $scope.appState = new AppState(getExplorerDefaultAppState());
+    const { mlExplorerFilter, mlExplorerSwimlane } = $scope.appState;
+
+    // Pass the current URL AppState on to anomaly explorer's reactive state.
+    // After this hand-off, the appState stored in explorerState$ is the single
+    // source of truth.
+    explorerAction$.next({
+      type: EXPLORER_ACTION.APP_STATE_SET,
+      payload: { mlExplorerSwimlane, mlExplorerFilter },
+    });
+
+    // This is temporary and can be removed once explorer.js migrated fully
+    // from explorer$ to explorerState$. This needs to be done only once
+    // the original URL AppState has been passed on to the observable state above.
+    explorerAction$.next(null);
+
+    // Now that appState in explorerState$ is the single source of truth,
+    // subscribe to it and update the actual URL appState on changes.
+    subscriptions.add(
+      explorerAppState$.subscribe((appState: IAppState) => {
+        $scope.appState.fetch();
+        $scope.appState = merge($scope.appState, appState);
+        $scope.appState.save();
+        $scope.$applyAsync();
+      })
+    );
+
+    subscriptions.add(
+      subscribeAppStateToObservable(AppState, 'mlShowCharts', showCharts$, () =>
+        $rootScope.$applyAsync()
+      )
+    );
+    subscriptions.add(
+      subscribeAppStateToObservable(AppState, 'mlSelectInterval', interval$, () =>
+        $rootScope.$applyAsync()
+      )
+    );
+    subscriptions.add(
+      subscribeAppStateToObservable(AppState, 'mlSelectSeverity', severity$, () =>
+        $rootScope.$applyAsync()
+      )
+    );
```

Comment:
you can use `merge` to combine these observables because they have the same subscription callback

Reply:
I looked into it a bit but I'm afraid it won't work here as is because of what the `subscribeAppStateToObservable` utility expects. We'll pick up work on `AppState` soon after this with migrating to new platform so we can have another look then. Hope to get rid of that whole section at one point :)